source/mqtt_client_connection.c

```diff
@@ -227,7 +271,17 @@ static void s_on_connection_resumed(
     }
 
     /* Ensure that python class is still alive */
-    PyObject *self = PyWeakref_GetObject(py_connection->self_proxy); /* borrowed reference */
+    PyObject *self = Py_None;
+#if PY_VERSION_HEX >= 0x030D0000                                  /* Check if Python version is 3.13 or higher */
+    if (PyWeakref_GetRef(py_connection->self_proxy, &self) < 0) { /* strong reference */
+        PyErr_WriteUnraisable(PyErr_Occurred());
+        goto on_done;
+    }
+#else
+    /* PyWeakref_GetObject is deprecated since python 3.13 */
+    self = PyWeakref_GetObject(py_connection->self_proxy); /* borrowed reference */
+#endif
```

Comment:
the logic is copy/pasted a lot of time. How about backporting PyWeakref_GetRef to older versions

module.h:
```
#if PY_VERSION_HEX < 0x30D0000
// PyWeakref_GetRef() was added in Python3.13, make it available in older versions
int PyWeakref_GetRef(PyObject *ref, PyObject **pobj);
#endif
```

module.c:
```
#if PY_VERSION_HEX < 0x30D0000
// PyWeakref_GetRef() was added in Python3.13, make it available in older versions
int PyWeakref_GetRef(PyObject *ref, PyObject **pobj) {
    ...
}
#endif
```

Reply:
https://github.com/python/cpython/blob/a9c2bc16349c2be3005f97249f3ae9699988f218/Objects/weakrefobject.c#L931-L947

Reply:
Not sure if we can backport it. `PyWeakref_GetRef` depends upon the internal function `_PyWeakref_GetRef`, which is probably located at https://fossies.org/linux/misc/Python-3.13.0rc2.tar.xz/Python-3.13.0rc2/Include/internal/pycore_weakref.h, and that function also depends upon several internal functions.

We can potentially add `PyWeakref_GetRef`, which calls `PyWeakref_GetObject` and also increments the ref count, so that we can avoid the copy-paste.

Reply:
Thanks, updated. I have created a new function which abstracts the differences between APIs into a single place, and we don't need to do any code duplication. I didn't define `PyWeakref_GetRef` for pre-Python 3.13 because I thought it would be confusing for future readers that we are calling the Python 3.13+ function, and having our own definition for pre-Python 3.13 which does something slightly different.